target/linux/ath79/files/drivers/mtd/nand/raw/rb4xx-nand.c

```diff
@@ -0,0 +1,302 @@
+	.free = rb4xx_ooblayout_free,
+};
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,20,0)
+static uint8_t rb4xx_nand_read_byte(struct mtd_info *mtd)
+{
+	struct rb4xx_nand *nand = mtd->priv;
+#else
+static u8 rb4xx_nand_read_byte(struct nand_chip *chip)
+{
+	struct rb4xx_nand *nand = chip->priv;
+#endif
+	struct rb4xx_cpld *cpld = nand->cpld;
+	u8 data;
+	int ret;
+
+	ret = cpld->rb4xx_cpld_read_nand(cpld, &data, 1);
+	if (unlikely(ret))
+		return 0xff;
+
+	return data;
+}
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,20,0)
+static void rb4xx_nand_write_buf(struct mtd_info *mtd, const uint8_t *buf,
+				 int len)
+{
+	struct rb4xx_nand *nand = mtd->priv;
+#else
+static void rb4xx_nand_write_buf(struct nand_chip *chip, const u8 *buf, int len)
+{
+	struct rb4xx_nand *nand = chip->priv;
+#endif
+	struct rb4xx_cpld *cpld = nand->cpld;
+
+	cpld->rb4xx_cpld_write_nand(cpld, buf, len);
+}
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,20,0)
+static void rb4xx_nand_read_buf(struct mtd_info *mtd, uint8_t *buf, int len)
+{
+	struct rb4xx_nand *nand = mtd->priv;
+#else
+static void rb4xx_nand_read_buf(struct nand_chip *chip, u8 *buf, int len)
+{
+	struct rb4xx_nand *nand = chip->priv;
+#endif
+	struct rb4xx_cpld *cpld = nand->cpld;
+
+	cpld->rb4xx_cpld_read_nand(cpld, buf, len);
+}
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,20,0)
+static void rb4xx_nand_cmd_ctrl(struct mtd_info *mtd, int dat, unsigned int ctrl)
+{
+	struct rb4xx_nand *nand = mtd->priv;
+#else
+static void rb4xx_nand_cmd_ctrl(struct nand_chip *chip, int dat,
+				unsigned int ctrl)
+{
+	struct rb4xx_nand *nand = chip->priv;
+#endif
+	struct rb4xx_cpld *cpld = nand->cpld;
+	u8 data = dat;
+
+	if (ctrl & NAND_CTRL_CHANGE) {
+		gpiod_set_value_cansleep(nand->cle, !!(ctrl & NAND_CLE));
+		gpiod_set_value_cansleep(nand->ale, !!(ctrl & NAND_ALE));
+		gpiod_set_value_cansleep(nand->nce,  !(ctrl & NAND_NCE));
+	}
+
+	if (dat != NAND_CMD_NONE)
+		cpld->rb4xx_cpld_write_nand(cpld, &data, 1);
+}
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(4,20,0)
+static int rb4xx_nand_dev_ready(struct mtd_info *mtd)
+{
+	struct rb4xx_nand *nand = mtd->priv;
+#else
+static int rb4xx_nand_dev_ready(struct nand_chip *chip)
+{
+	struct rb4xx_nand *nand = chip->priv;
+#endif
+
+	return gpiod_get_value_cansleep(nand->rdy);
+}
+
+static int rb4xx_nand_probe(struct platform_device *pdev)
+{
+	struct device *dev = &pdev->dev;
+	struct device *parent = dev->parent;
+	struct rb4xx_nand *nand;
+	struct mtd_info *mtd;
+	int ret;
+
+	if (!parent)
+		return -ENODEV;
+
+	/* TODO fix the "DMA mask not set" warnings?
+	 * see conversation here https://lkml.org/lkml/2020/4/28/782
+	 *
+	 * perhaps: of_dma_configure(dev, NULL, true);
+	 */
+
+	nand = devm_kzalloc(dev, sizeof(*nand), GFP_KERNEL);
+	if (!nand)
+		return -ENOMEM;
+
+	platform_set_drvdata(pdev, nand);
+	nand->cpld	= dev_get_drvdata(parent);
+	nand->dev	= dev;
+
+	/* TODO possible to use the GPIO names?
+	 * May need to defer the probe() for them to become available?
+	 */
```

Comment:
After rereading the DTS it makes sense that you can't use the names: they're not defined for the nand driver. Using index is therefore correct. The only expectation is that the DTS must provide them in the right order.

Reply:
Do you think it's worth leaving these names in the DTS for documentation purposes, or should they be deleted?

Reply:
I think the DTS looks good as it is. The comment in the nand gpios make it very clear what they are and so comparing with the code is easy. I'd leave them.